### ISSUE TITLE
Race conditions when using Threads

### ISSUE THREAD

**Issue by dr0i** [metafacture/metafacture-fix#255]
Came up in https://github.com/hbz/lobid-resources/issues/1409#issuecomment-1235382923.
This one is not reproducible and seems to be the result of Metafix not being thread safe:

> WARN [Thread-4] (Metafix.java:421) - Error while executing Fix expression (at file:/home/sol/git/lobid-resources-alma-fix/src/main/resources/alma/fix/contribution.fix, line 184): do list(path:"110[012] ", "var":"$i") unless exists("$i.M") unless exists("$i.a.1") unless exists("$i.4") if any_match("type[]","ArchivedWebPage|Miscellaneous|Bibliography|Statistics|Legislation|PublishedScore|Game|Image|Map|Standard") add_field("$i.4","cre") elsif any_match("type[]","Periodical|Collection|Series|Newspaper|Journal|PublicationIssue|EditedVolume|Proceedings|Festschrift|ReferenceSource") add_field("$i.4","edt") elsif any_match("type[]","Book|MultiVolumeBook|Article|Thesis|OfficialPublication|Schoolbook|Biography|Report") add_field("$i.4","aut") else add_field("$i.4","cre") end end do list(path: "$i.4", "var":"$j") set_hash("contribution[].$append.agent") do list(path:"$i.0","var":"$k") if all_match("$k", "^\\(DE-588\\).*$") paste("contribution[].$last.agent.gndIdentifier","$k") paste("contribution[].$last.agent.id","$k") end end paste("contribution[].$last.agent.label", "$i.a", "$i.b", join_char: ". ") set_array("contribution[].$last.agent.type[]","CorporateBody") copy_field("$j","contribution[].$last.role.id") end end end end
java.util.ConcurrentModificationException
        at java.util.HashMap$HashIterator.nextNode(HashMap.java:1469)
        at java.util.HashMap$KeyIterator.next(HashMap.java:1493)
        at org.metafacture.commons.tries.WildcardTrie.matches(WildcardTrie.java:128)
        at org.metafacture.commons.tries.WildcardTrie.get(WildcardTrie.java:123)
        at org.metafacture.commons.tries.SimpleRegexTrie.get(SimpleRegexTrie.java:75)
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:806)
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:778)
        at org.metafacture.metafix.Value$Hash.containsField(Value.java:522)
        at org.metafacture.metafix.Record.get(Record.java:124)
        at org.metafacture.metafix.FixBind$1.execute(FixBind.java:33)
        at org.metafacture.metafix.RecordTransformer.lambda$null$4(RecordTransformer.java:137)
        at org.metafacture.metafix.RecordTransformer.lambda$null$22(RecordTransformer.java:225)
        at org.metafacture.metafix.RecordTransformer.lambda$null$1(RecordTransformer.java:124)
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:236)
        at org.metafacture.metafix.RecordTransformer.lambda$transform$2(RecordTransformer.java:124)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:123)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:115)
        at org.metafacture.metafix.FixMethod$1.apply(FixMethod.java:60)
        at org.metafacture.metafix.RecordTransformer.lambda$null$20(RecordTransformer.java:205)
        at org.metafacture.metafix.RecordTransformer.lambda$null$22(RecordTransformer.java:225)
        at org.metafacture.metafix.RecordTransformer.lambda$null$1(RecordTransformer.java:124)
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:236)
        at org.metafacture.metafix.RecordTransformer.lambda$transform$2(RecordTransformer.java:124)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:123)
        at org.metafacture.metafix.Metafix.endRecord(Metafix.java:202)
        at org.metafacture.biblio.marc21.MarcXmlHandler.endElement(MarcXmlHandler.java:132)
        at org.apache.xerces.parsers.AbstractSAXParser.endElement(Unknown Source)
        at org.apache.xerces.impl.XMLNSDocumentScannerImpl.scanEndElement(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source)
        at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
        at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
        at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:69)
        at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:43)
        at org.metafacture.strings.StringReader.process(StringReader.java:48)
        at org.metafacture.strings.StringReader.process(StringReader.java:34)
        at org.metafacture.flowcontrol.ObjectPipeDecoupler$Feeder.run(ObjectPipeDecoupler.java:175)
        at java.lang.Thread.run(Thread.java:748)

The threads are created using [Metafacture's flowcontrol.ObjectPipeDecoupler](https://github.com/metafacture/metafacture-core/blob/master/metafacture-flowcontrol/src/main/java/org/metafacture/flowcontrol/ObjectPipeDecoupler.java).

**Comment by dr0i** [comment]
From https://github.com/hbz/lobid-resources/issues/1409#issuecomment-1235524415 , might also be a result of unsafe threads:
>  INFO [Thread-5] (Metafix.java:414) - Current record: {leader=01879ngm a2200385ua 4500, 006=m     o  c        , 007=[cr una---unuuu, vz uzazuu], 008=160826p20162006cau023        o   vlund d, 001=991025930286006442, 02852={a=1089373, b=Kanopy}, 035  =[{a=(OCoLC)956893319, 0=(uri) http://www.worldcat.org/oclc/956893319}, {a=(CaSfKAN)kan1089373}], 040  ={a=VDU, b=eng, c=VDU}, 24500={a=Lockie Leonard Series 1. Episode 13: It's not You, it's Me}, 264 1=[{b=Australian Children's Television Foundation,, c=2006.}, {a=[San Francisco, California, USA] :, b=Kanopy Streaming,, c=2016.}], 300  ={a=1 online resource (streaming video file) (24 minutes):, b=digital, .flv file, sound}, 306  ={a=002354}, 336  ={a=tdi, b=tdi, 2=rdacontent}, 337  ={b=[c, c], 2=rdamedia}, 338  ={b=[cr, cr], 2=rdacarrier}, 500  =[{a=Filmclip}, {a=Title from title frames.}, {a=In Process Record.}], 518  ={a=Originally produced by Australian Children's Television Foundation in 2006.}, 520  ={a=Lockie has some problems.  First, he's decided to break up with Dot.  It's just figuring out how that's tricky.  Meanwhile, Mum is acting weird.  She's been staring at the TV for days, when it's not even tuned to a channel.  Phillip's not impressed - he's missing his favourite superhero cartoon, Captain Chicken.   And to top things off, Egg and Lockie's quest to save the river is going nowhere...    Lockie has got some serious work to do.   And he's going to have to find his inner Captain Chicken to sort it all out.}, 538  ={a=Mode of access: World Wide Web.}, 546  ={a=In English}, 653  ={a=Australian Studies}, 7102 =[{a=Australian Children's Television Foundation (Firm),, 4=dst}, {a=Kanopy (Firm),, 4=dst}], 85640={u=https://bielefeld.kanopy.com/node/89374, z=A Kanopy streaming video}, 85642={z=Cover Image, u=https://www.kanopy.com/node/89374/external-image}, 9801 ={a=ZDB-177-KAN}, MBD  ={M=49HBZ_BIE, i=991025930286006442, n=Universität Bielefeld}, MNG  ={c=System, f=OTHER, i=marc21, k=01, e=false, d=2022-06-07 18:14:09 Europe/Berlin, g=kan1089373, j=00, a=System, b=2022-06-07 18:14:11 Europe/Berlin}, POR  ={x=System, a=53344020580006442, M=49HBZ_BIE, p=61334816600006442, q=Kanopy, c=static, B=17.131 Titel, l=A Kanopy streaming video, y=2022-06-07 18:14:20 Europe/Berlin, w=2022-06-07 18:14:16 Europe/Berlin, D=https://eu04.alma.exlibrisgroup.com/view/uresolver/49HBZ_NETWORK/openurl?u.ignore_date_coverage=true&portfolio_pid=53344020580006442&Force_direct=true, v=System, z=2022-06-07 16:14:16, i=false, g=UB_BI, t=BIE-8465, S=62343002520006442, d=https://hbz-network.userservices.exlibrisgroup.com/view/uresolver/49HBZ_NETWORK/openurl?u.ignore_date_coverage=true&rft.mms_id=991025930286006442, b=Available, f=VIDEO, e=https://bielefeld.kanopy.com/node/89374, 8=53344020580006442}, @context=http://lobid.org/resources/context.jsonld, almaMmsId=991025930286006442, id=http://lobid.org/resources/991025930286006442#!, urn[]=[]}
 WARN [Thread-4] (Metafix.java:421) - Error while executing Fix expression (at file:/home/sol/git/lobid-resources-alma-fix/src/main/resources/alma/fix/identifiers.fix, line 23): do list(path: "0247?", "var": "$i") if any_equal("$i.2","urn") copy_field("$i.a","urn[].$append") end end
java.lang.IllegalStateException: Key '0' already used
        at org.metafacture.commons.tries.CharMap.put(CharMap.java:112)
        at org.metafacture.commons.tries.CharMap.put(CharMap.java:97)
        at org.metafacture.commons.tries.WildcardTrie$Node.addNext(WildcardTrie.java:158)
        at org.metafacture.commons.tries.WildcardTrie.simplyPut(WildcardTrie.java:79)
        at org.metafacture.commons.tries.WildcardTrie.put(WildcardTrie.java:67)
        at org.metafacture.commons.tries.SimpleRegexTrie.put(SimpleRegexTrie.java:63)
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:789)
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:778)
        at org.metafacture.metafix.Value$Hash.containsField(Value.java:522)
        at org.metafacture.metafix.Record.get(Record.java:124)
        at org.metafacture.metafix.FixBind$1.execute(FixBind.java:33)
        at org.metafacture.metafix.RecordTransformer.lambda$null$4(RecordTransformer.java:137)
        at org.metafacture.metafix.RecordTransformer.lambda$null$22(RecordTransformer.java:225)
        at org.metafacture.metafix.RecordTransformer.lambda$null$1(RecordTransformer.java:124)
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:236)
        at org.metafacture.metafix.RecordTransformer.lambda$transform$2(RecordTransformer.java:124)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:123)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:115)
        at org.metafacture.metafix.FixMethod$1.apply(FixMethod.java:60)
        at org.metafacture.metafix.RecordTransformer.lambda$null$20(RecordTransformer.java:205)
        at org.metafacture.metafix.RecordTransformer.lambda$null$22(RecordTransformer.java:225)
        at org.metafacture.metafix.RecordTransformer.lambda$null$1(RecordTransformer.java:124)
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:236)
        at org.metafacture.metafix.RecordTransformer.lambda$transform$2(RecordTransformer.java:124)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:123)
        at org.metafacture.metafix.Metafix.endRecord(Metafix.java:202)
        at org.metafacture.biblio.marc21.MarcXmlHandler.endElement(MarcXmlHandler.java:132)
        at org.apache.xerces.parsers.AbstractSAXParser.endElement(Unknown Source)
        at org.apache.xerces.impl.XMLNSDocumentScannerImpl.scanEndElement(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source)
        at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
        at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
        at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:69)
        at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:43)
        at org.metafacture.strings.StringReader.process(StringReader.java:48)
        at org.metafacture.strings.StringReader.process(StringReader.java:34)
        at org.metafacture.flowcontrol.ObjectPipeDecoupler$Feeder.run(ObjectPipeDecoupler.java:175)
        at java.lang.Thread.run(Thread.java:748)

**Comment by dr0i** [comment]
Some more exceptions - they were plentiful and this is all I have left in my notes:
a)
> java.lang.NullPointerException
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:813)
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:778)
        at org.metafacture.metafix.Value$Hash.containsField(Value.java:522)
        at org.metafacture.metafix.Record.get(Record.java:124)
        at org.metafacture.metafix.FixBind$1.execute(FixBind.java:33)

b)
 >   at java.util.regex.Matcher.start(Matcher.java:343)
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:792)
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:778)
        at org.metafacture.metafix.Value$Hash.containsField(Value.java:522)
        at org.metafacture.metafix.Record.get(Record.java:124)

c)
> java.lang.StringIndexOutOfBoundsException: String index out of range: 6
        at java.lang.String.substring(String.java:1963)
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:792)
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:778)
        at org.metafacture.metafix.Value$Hash.containsField(Value.java:522)
        at org.metafacture.metafix.Record.get(Record.java:124)

d)
> java.util.ConcurrentModificationException
        at java.util.HashMap$HashIterator.nextNode(HashMap.java:1469)
        at java.util.HashMap$KeyIterator.next(HashMap.java:1493)
        at org.metafacture.commons.tries.WildcardTrie.matches(WildcardTrie.java:128)
        at org.metafacture.commons.tries.WildcardTrie.get(WildcardTrie.java:123)
        at org.metafacture.commons.tries.SimpleRegexTrie.get(SimpleRegexTrie.java:75)
        at org.metafacture.metafix.Value$Hash.findFields(Value.java:806)

e)
>  at org.metafacture.metafix.FixPath.insertInto(FixPath.java:297)
        at org.metafacture.metafix.FixPath.insertInto(FixPath.java:263)
        at org.metafacture.metafix.FixPath.lambda$null$20(FixPath.java:292)
        at org.metafacture.metafix.Value$TypeMatcher.match(Value.java:349)
        at org.metafacture.metafix.Value$TypeMatcher.ifArray(Value.java:322)

**Comment by blackwinter** [comment]
I assume the exception in case b) was `java.lang.IllegalStateException` and in case e) it was `java.lang.IllegalArgumentException`, right?

**Comment by dr0i** [comment]
Yep, right!

**Comment by blackwinter** [comment]
Case e) seems to be a data/Fix issue, maybe you can show the full error message (including the current record).

Everything else boils down to `Value$Hash.findFields()`, which is also supported by the analysis that the only `static` mutable, non-thread-safe objects are used there (`PREFIX_CACHE` = `HashMap`, `TRIE_CACHE` = `HashMap`, `TRIE` = `SimpleRegexTrie`). So my hypothesis would be that it's sufficient to make this one method `synchronized` (individually thread-safe data structures are not sufficient as these collections depend on each other, they have to be collectively thread-safe).

Now I would definitely like to create tests for these thread-safety issues but I haven't gotten anywhere yet.[^1] Maybe you could just do some manual testing in order to verify the fix. We should also bear in mind that this method is absolutely on the critical path (cf. #207) so we definitely have to keep an eye on how it affects performance.

[^1]: Some pointers for thread-safety testing: [jcstress](https://github.com/openjdk/jcstress), [ConcurrentUnit](https://github.com/jhalterman/concurrentunit), [MultithreadedTC](https://www.cs.umd.edu/projects/PL/multithreadedtc/), [vmlens](https://vmlens.com/)

**Comment by blackwinter** [comment]
Benchmark results:

| Benchmark   | (fixDef) | (input)    |   Score @ c8ba171 |              Score @ static | Units   |  Boost |
|:------------|:---------|:-----------|------------------:|----------------------------:|:--------|-------:|
| Baseline    | N/A      | N/A        |  3031.762 ± 4.847 |            3031.196 ± 4.872 | ops/us  | -0.02% |
| FixParse    | nothing  | N/A        |    34.818 ± 0.728 |              34.655 ± 0.906 | ops/s   | -0.47% |
| FixParse    | alma     | N/A        |    25.210 ± 0.681 |              25.225 ± 0.732 | ops/s   | +0.06% |
| Metafix     | nothing  | empty      | 2263.035 ± 60.509 |           2244.932 ± 19.264 | ops/s   | -0.80% |
| Metafix     | nothing  | alma-small |    38.668 ± 1.142 |              37.977 ± 1.920 | ops/s   | -1.79% |
| Metafix     | alma     | empty      |  767.576 ± 10.739 |             764.330 ± 1.974 | ops/s   | -0.42% |
| Metafix     | alma     | alma-small |     4.042 ± 0.046 |               3.981 ± 0.037 | ops/s   | -1.51% |
| SlowMetafix | nothing  | alma-large |    23.015 ± 0.994 |              23.037 ± 0.804 | ops/min | +0.10% |
| SlowMetafix | alma     | alma-large |     2.474 ± 0.036 |               2.462 ± 0.032 | ops/min | -0.49% |
                                                                                                               
| Benchmark   | (fixDef) | (input)    |   Score @ c8ba171 | Score @ static+synchronized | Units   |  Boost |
|:------------|:---------|:-----------|------------------:|----------------------------:|:--------|-------:|
| Baseline    | N/A      | N/A        |  3031.762 ± 4.847 |            3031.383 ± 1.551 | ops/us  | -0.01% |
| FixParse    | nothing  | N/A        |    34.818 ± 0.728 |              34.834 ± 0.871 | ops/s   | +0.05% |
| FixParse    | alma     | N/A        |    25.210 ± 0.681 |              25.480 ± 0.713 | ops/s   | +1.07% |
| Metafix     | nothing  | empty      | 2263.035 ± 60.509 |           2256.579 ± 26.081 | ops/s   | -0.29% |
| Metafix     | nothing  | alma-small |    38.668 ± 1.142 |              38.282 ± 0.188 | ops/s   | -1.00% |
| Metafix     | alma     | empty      |  767.576 ± 10.739 |             782.327 ± 8.896 | ops/s   | +1.92% |
| Metafix     | alma     | alma-small |     4.042 ± 0.046 |               3.956 ± 0.042 | ops/s   | -2.13% |
| SlowMetafix | nothing  | alma-large |    23.015 ± 0.994 |              22.901 ± 0.322 | ops/min | -0.50% |
| SlowMetafix | alma     | alma-large |     2.474 ± 0.036 |               2.440 ± 0.059 | ops/min | -1.37% |

**Comment by blackwinter** [comment]
Benchmark results for configurable synchronization (intrinsic lock):

<details><summary>Details</summary>

```diff
@@ -504,6 +504,8 @@ public class Value implements JsonValue { // checkstyle-disable-line ClassDataAb
         private static final Map<String, Map<String, Boolean>> TRIE_CACHE = new HashMap<>();
         private static final SimpleRegexTrie<String> TRIE = new SimpleRegexTrie<>();
 
+        private static final Object LOCK = Boolean.getBoolean("org.metafacture.metafix.threadSafe") ? new Object() : null;
+
         private final Map<String, Value> map = new LinkedHashMap<>();
 
         /**
@@ -775,7 +777,14 @@ public class Value implements JsonValue { // checkstyle-disable-line ClassDataAb
             final Set<String> fieldSet = new LinkedHashSet<>();
 
             for (final String term : ALTERNATION_PATTERN.split(pattern)) {
-                findFields(term, fieldSet);
+                if (LOCK != null) {
+                    synchronized (LOCK) {
+                        findFields(term, fieldSet);
+                    }
+                }
+                else {
+                    findFields(term, fieldSet);
+                }
             }
```

Set `-Dorg.metafacture.metafix.threadSafe=true` to enable lock.
</details>

| Benchmark   | (fixDef) | (input)    |   Score @ c8ba171 | Score @ lock-disabled | Units   |  Boost |
|:------------|:---------|:-----------|------------------:|----------------------:|:--------|-------:|
| Baseline    | N/A      | N/A        |  3031.762 ± 4.847 |      3031.914 ± 3.001 | ops/us  | +0.01% |
| FixParse    | nothing  | N/A        |    34.818 ± 0.728 |        34.599 ± 0.754 | ops/s   | -0.63% |
| FixParse    | alma     | N/A        |    25.210 ± 0.681 |        25.316 ± 0.653 | ops/s   | +0.42% |
| Metafix     | nothing  | empty      | 2263.035 ± 60.509 |     2261.131 ± 28.884 | ops/s   | -0.08% |
| Metafix     | nothing  | alma-small |    38.668 ± 1.142 |        40.038 ± 0.282 | ops/s   | +3.54% |
| Metafix     | alma     | empty      |  767.576 ± 10.739 |       781.751 ± 1.895 | ops/s   | +1.85% |
| Metafix     | alma     | alma-small |     4.042 ± 0.046 |         3.976 ± 0.284 | ops/s   | -1.63% |
| SlowMetafix | nothing  | alma-large |    23.015 ± 0.994 |        22.976 ± 0.860 | ops/min | -0.17% |
| SlowMetafix | alma     | alma-large |     2.474 ± 0.036 |         2.490 ± 0.012 | ops/min | +0.65% |

| Benchmark   | (fixDef) | (input)    |       Score @ c8ba171 |  Score @ lock-enabled | Units   |  Boost |
|:------------|:---------|:-----------|----------------------:|----------------------:|:--------|-------:|
| Baseline    | N/A      | N/A        |      3031.762 ± 4.847 |      3030.142 ± 7.040 | ops/us  | -0.05% |
| FixParse    | nothing  | N/A        |        34.818 ± 0.728 |        34.218 ± 0.774 | ops/s   | -1.72% |
| FixParse    | alma     | N/A        |        25.210 ± 0.681 |        25.353 ± 0.840 | ops/s   | +0.57% |
| Metafix     | nothing  | empty      |     2263.035 ± 60.509 |     2227.026 ± 15.194 | ops/s   | -1.59% |
| Metafix     | nothing  | alma-small |        38.668 ± 1.142 |        38.913 ± 0.441 | ops/s   | +0.63% |
| Metafix     | alma     | empty      |      767.576 ± 10.739 |      756.706 ± 15.609 | ops/s   | -1.42% |
| Metafix     | alma     | alma-small |         4.042 ± 0.046 |         4.063 ± 0.079 | ops/s   | +0.52% |
| SlowMetafix | nothing  | alma-large |        23.015 ± 0.994 |        23.112 ± 0.149 | ops/min | +0.42% |
| SlowMetafix | alma     | alma-large |         2.474 ± 0.036 |         2.435 ± 0.007 | ops/min | -1.58% |

**Comment by dr0i** [comment]
Updated Case e).
Benchmarking looks good I think.
Going to test with 20M records. Can you open a branch for convenient testing?

**Comment by blackwinter** [comment]
> Can you open a branch for convenient testing?

For which implementation? They're both pretty straight forward, though, and it depends on which one you want to test.

**Comment by dr0i** [comment]
Would want to test both.
Concerning the configurable one:
I opt for "threadsafe" as default and setting to "unsafe" optinonal.

**Comment by blackwinter** [comment]
Pushed both branches.

> I opt for "threadsafe" as default and setting to "unsafe" optionally.

Sure, we can decide it later. I've stuck to the simpler implementation for now.

**Comment by blackwinter** [comment]
> Updated Case e).

That's indeed an unrelated issue. `label` is an array, so `replace_all()` fails.

**Comment by dr0i** [comment]
Tried to test `255-makeFindFieldsOptionallyThreadSafe` but forgot to set the option :P .
The outcome, however, resulted in zero errors. Also remarkable: all 24M records needed only 4h (!) ( executed with 6 threads on a 12 core server with near to zero load (the latter may explain the speed). Java options were `-Xmx8G -Xms2G`).
Now, it's hard to reproduce unsafe thread exceptions as we all know. But I wonder if https://github.com/metafacture/metafacture-fix/pull/259 has resolved these btw resp. made these so much more unlikely to happen that they didn't happen (this time).
I am going to test your branch with option set to see the performance impact. We would then run the ETL some time without that option and see if at some point thread induced exceptions appear and heal them with the option set (if the performance test gives reasonable results).

**Comment by dr0i** [comment]
Dump my last comment: configuration was wrong, it used the morph :(

**Comment by blackwinter** [comment]
Limetrans results for writing approx. 2m records to local file:

| Metafix                                                                                 | Runtime  | Boost  |
|:----------------------------------------------------------------------------------------|:---------|:-------|
| `master` (98bd5cd)                                                                      | 4h59m38s |        |
| `255-makeFindFieldsThreadSafe` (78b985b)                                                | 5h01m10s | +0.51% |
| `255-makeFindFieldsOptionallyThreadSafe` (ad7d84e)                                      | 4h58m42s | -0.31% |
| `255-makeFindFieldsOptionallyThreadSafe` w/ `-Dorg.metafacture.metafix.threadSafe=true` | 5h04m42s | +1.69% |

**Comment by dr0i** [comment]
Did the test with 24M docs. Results are good in the way as no more errors occurred, but disappointing as the needed time  rose from ~9h to ~25h. Double checked that. I used https://github.com/metafacture/metafacture-fix/commit/ad7d84e9d0c886fedc55a311cb18634d4d69a19a.

**Comment by blackwinter** [comment]
> Results are good in the way as no more errors occurred,

Great!

> but disappointing as the needed time rose from ~9h to ~25h.

That's Morph vs. Fix, right? That was to be expected, wasn't it? It's not even as bad as I had anticipated ;)

**Comment by dr0i** [comment]
> That's Morph vs. Fix, right? 

no, it's _both_ using Fix.

**Comment by blackwinter** [comment]
> no, it's _both_ using Fix.

Ah, okay. Must be thread contention then. Maybe reducing the number of threads helps?

**Comment by dr0i** [comment]
Funny that you propose _decreasing_ of threads while I would assume that _increasing_ them would gain us speed.
(The execution is done using 6 Threads (while it's an 12-Core CPU where we could safely use max 11 threads).
Gonna test both scenarios.
> (You're probably not going to get it down to the unsafe runtime,

sure, but an increase of factor 2.5 feels just bad.  Also, as I have stated in https://github.com/metafacture/metafacture-fix/pull/257#issuecomment-1249233831, we _may_ just want to go with thread-unsafe and take the loss of 100 docs or so? Ok, this maybe even feels bader ;)

**Comment by blackwinter** [comment]
> Funny that you propose _decreasing_ of threads while I would assume that _increasing_ them would gain us speed.

Well, if the threads are fighting over a common resource, I would assume that increasing the number of contestants can only make it worse. But you're right, data over speculation ;)

> we _may_ just want to go with thread-unsafe and take the loss of 100 docs or so? Ok, this maybe even feels bader ;)

That would be an absolute no-go for us at least. Sure, some hundred(s) of documents out of tens of millions might never be noticed by users, but I think we should aim for completeness as much as possible for the union catalogue (especially as it would even be non-deterministic).

**Comment by blackwinter** [comment]
> I am not sure if it just bails out or corruption takes place. _If_ it ignores the concurrent modificated docs we could just reindex these, but am I right that some of the errors indicates that corruption takes place?

Whether documents with exceptions are ignored depends on your `strictnessHandlesProcessExceptions` and `strictness` settings. With your current setup, only the particular Fix expression is skipped. So there might be a field missing or it might not be properly formatted or something like that.

But we don't know if all error cases lead to exceptions, do we? Can there be silent corruption with unsynchronized accesses?

(Can you compare your safe and unsafe indexing results to determine the potential types and extent of corruption?)

**Comment by dr0i** [comment]
> Can there be silent corruption with unsynchronized accesses 

thinking this over: definitely yes (my proposal was a stupid idea )  => will not compare safe and unsafe indexing

> Gonna test both scenarios. 

Tests always in thread-safe mode. Values approximate. Also, theses values may vary from time to time (and I don't know why. Will eventually have more data and provide ranges instead of hard numbers):

3 threads: 100k/20m/thread => 26h/24M 
6 threads: 100k/30m/thread => 20h/24M
11 threads:  dumped that, CPU usage is stuck around 600%

Interestingly:  11 threads keeps the CPU usage stuck around 600% (instead of the optimal 1100%). So there is _definitely_ a limit where threads are making sense here (in the context of this server). I.e., it is as you said, at some point adding threads only increases the overhead without using allocated CPU cycles.

Checking also atm if using https://github.com/metafacture/metafacture-fix/commit/78b985be2e3597a71001b6769a2310719994b87f makes a difference.

Maybe it makes more sense to use gnu parallel to split the BGZF and start several single threaded JVMs. (I read somewhere that starting new JVMs from within a JVM is only possible for java >=10)

**Comment by blackwinter** [comment]
An alternative approach would be to move the problematic parts into the `Metafix` _instance_, then we wouldn't need any synchronization. But it would require extensive changes throughout the code.

**Comment by dr0i** [comment]
78b985be2e3597a71001b6769a2310719994b87f needed  27h  for the complete ETL (7 threads) :( .
Gonna try again https://github.com/metafacture/metafacture-fix/commit/ad7d84e9d0c886fedc55a311cb18634d4d69a19a , which needed 25h as stated in https://github.com/metafacture/metafacture-fix/issues/255#issuecomment-1271326838, but will this time use 7 threads instead of 6. (in general speaking, both thread-safe approaches seem not to make a difference).

**Comment by blackwinter** [comment]
Another alternative might be to utilise thread-local variables (see 662a3b9).

**Comment by dr0i** [comment]
Your `ThreadLocal` approach seems very promising, as it uses 100% CPU /Thread!  (just discarded the (redundant, not promising looking) test mentioned in https://github.com/metafacture/metafacture-fix/issues/255#issuecomment-1273097215).

First 100k needed only:
7 threads: 100k/20m/thread => ~12h/24M
Gonna test using 10 threads. Results tomorrow.

**Comment by blackwinter** [comment]
Benchmark results for thread-local variables:

| Benchmark   | (fixDef) | (input)    |   Score @ c8ba171 |   Score @ 662a3b9 | Units   |  Boost |
|:------------|:---------|:-----------|------------------:|------------------:|:--------|-------:|
| Baseline    | N/A      | N/A        |  3031.762 ± 4.847 |  3031.059 ± 4.275 | ops/us  | -0.02% |
| FixParse    | nothing  | N/A        |    34.818 ± 0.728 |    34.787 ± 0.829 | ops/s   | -0.09% |
| FixParse    | alma     | N/A        |    25.210 ± 0.681 |    25.381 ± 0.591 | ops/s   | +0.68% |
| Metafix     | nothing  | empty      | 2263.035 ± 60.509 | 2263.331 ± 57.177 | ops/s   | +0.01% |
| Metafix     | nothing  | alma-small |    38.668 ± 1.142 |    37.573 ± 0.949 | ops/s   | -2.83% |
| Metafix     | alma     | empty      |  767.576 ± 10.739 |  785.372 ± 15.244 | ops/s   | +2.32% |
| Metafix     | alma     | alma-small |     4.042 ± 0.046 |     4.039 ± 0.037 | ops/s   | -0.07% |
| SlowMetafix | nothing  | alma-large |    23.015 ± 0.994 |    22.666 ± 1.245 | ops/min | -1.52% |
| SlowMetafix | alma     | alma-large |     2.474 ± 0.036 |     2.452 ± 0.061 | ops/min | -0.89% |

**Comment by dr0i** [comment]
`ThreadLocal ` completes ETL 24M Alma Fix with 10 threads in 6h10m :+1: 

Suggestion: open PR against `255-makeFindFieldsThreadSafeWithThreadLocal` and merge into master. Close this issue.

**Comment by blackwinter** [comment]
Excellent :) Let me just verify that there's no regression in Limetrans.

I've opened #263 which you can now review; if everything works out I'll merge it tomorrow. Okay?

**Comment by dr0i** [comment]
> I've opened https://github.com/metafacture/metafacture-fix/pull/263 which you can now review; if everything works out I'll merge it tomorrow. Okay?

ahm. Did read this too late. Already merged. I reopen this issue so we can react properly if you find a regression.

**Comment by blackwinter** [comment]
Limetrans results for writing approx. 2m records to local file:

| Metafix                                                                                 | Runtime  | Boost  |
|:----------------------------------------------------------------------------------------|:---------|:-------|
| `master` (ce9157f)                                                                      | 4h36m49s |        |
| `255-makeFindFieldsThreadSafeWithThreadLocal` (0f2a0df)                                 | 4h25m05s | -4.24% |
